### PR TITLE
STS: add support for signing requests with `assume_role` tokens

### PR DESCRIPTION
**Pull request by zbaylin**
This supersedes #95.

It includes the ability to sign AWS requests with `assume_role` tokens, and has other various fixes, such as fixing the way dune test files are generated if extra libraries are needed.

### CODE REVIEW

File: async/runtime.mli [hunk 1]
```diff
@@ -35,6 +35,7 @@ val run_request :
      region:string
   -> access_key:string
   -> secret_key:string
+  -> ?token:string
```
Comment (tmcgilchrist) [inline async/runtime.mli:38]: Could you add documentation for `?token` into the odoc below?

File: libraries/s3/lib_test/test_async.ml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  let access_key = Unix.getenv "AWS_ACCESS_KEY_ID"
+
+  let secret_key = Unix.getenv "AWS_SECRET_ACCESS_KEY"
+
+  let token = None
+
+  let run_request = Aws_async.Runtime.run_request ~access_key ~secret_key ?token
+
+  let un_m v = Async.Thread_safe.block_on_async_exn (fun () -> v)
+end)
```
Comment (tmcgilchrist) [inline libraries/s3/lib_test/test_async.ml:15]: A couple of improvements:
1. I moved this logic into the `aws_service_test.ml` files so I could conditionally run the tests in CI (supporting both GitHub Actions and ocaml-ci) See https://github.com/inhabitedtype/ocaml-aws/blob/master/libraries/ec2/lib_test/aws_ec2_test.ml#L183-L200
1. the S3 bindings are broken and need some attention on the deserialisation https://github.com/inhabitedtype/ocaml-aws/issues/93

Could you remove the S3 generated code and move the test logic over as per aws_ec2 example, cheers.

File: libraries/sts/lib_test/aws_sts_test.ml [hunk 3]
```diff
@@ -67,10 +152,11 @@ functor
         try Some (Unix.getenv "AWS_SECRET_ACCESS_KEY") with Not_found -> None
       in
       let region = try Some (Unix.getenv "AWS_DEFAULT_REGION") with Not_found -> None in
+      let role_arn = try Some (Unix.getenv "AWS_STS_ROLE_ARN") with Not_found -> None in
 
       match access_key, secret_key, region with
       | Some access_key, Some secret_key, Some region ->
-          run_test_tt_main (suite { access_key; secret_key; region })
+          run_test_tt_main (suite { access_key; secret_key; region; role_arn })
```
Comment (tmcgilchrist) [inline libraries/sts/lib_test/aws_sts_test.ml:159]: Nice thanks for adding that test.

**Review by tmcgilchrist — CHANGES_REQUESTED** [review]
Overall the changes look good. A couple of things to fixup and could you add a line into CHANGES.md describing the work and link to this PR. 

Thank you for the contribution. :-)

File: libraries/s3/lib_test/test_async.ml (hunk 2 above)
Comment (zbaylin) [reply libraries/s3/lib_test/test_async.ml]: Hey @tmcgilchrist -- sorry for re-adding the broken S3 code. I removed the generated S3 test stuff from the branch.

I'm slightly confused about your second request. Do you want me to move the test logic for *other* libraries to be similar to the EC2 example you provided?

Let me know :)

File: async/runtime.mli (hunk 1 above)
Comment (zbaylin) [reply async/runtime.mli:38]: Should be added in 5e3e783 :)

File: libraries/s3/lib_test/test_async.ml (hunk 2 above)
Comment (tmcgilchrist) [reply libraries/s3/lib_test/test_async.ml]: Looks like it was just the S3 test that wasn't doing what I wanted. No need to do anything now.

**Review by tmcgilchrist — APPROVED** [review]

**Comment by tmcgilchrist** [comment]
Thanks @zbaylin @UnrealAkama @bleepbloopsify

**Pull request merged**